### PR TITLE
[ROCm] Add ROCm AMDGPU support for inductor cpp codegen

### PR DESCRIPTION
**Pull request by jataylo**
Follows from previous enablement attempt: https://github.com/pytorch/pytorch/pull/101797

Adds support for hsaco binaries in inductor's cpp_wrapper codegen and enables the CUDA tests in test_cpp_wrapper.

This PR also brings in additional required hipify mappings for the wrapper codegen file.

NOTE: we can unskip some of these tests when we enabled MI210 runners.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @hongxiayang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel

### CODE REVIEW

**Comment by jataylo** [comment]
**Example generated code with hipified cpp_wrapper**
```
import torch
from torch._inductor.codecache import CppWrapperCodeCache

cpp_wrapper_src = (
'''

#include <torch/csrc/inductor/inductor_ops.h>

#include <ATen/native/BinaryOps.h>
#include <c10/util/Exception.h>
#include <ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h>

#define AT_CUDA_DRIVER_CHECK_OVERRIDE(EXPR)                         \
do {                                                                \
    hipError_t __err = EXPR;                                          \
    if (__err != hipSuccess) {                                    \
        AT_ERROR("CUDA driver error: ", static_cast<int>(__err));   \
    }                                                               \
} while (0)

static inline hipFunction_t loadKernel(const std::string &filePath,
        const std::string &funcName) {
    hipModule_t mod;
    hipFunction_t func;
    AT_CUDA_DRIVER_CHECK_OVERRIDE(hipModuleLoad(&mod, filePath.c_str()));
    AT_CUDA_DRIVER_CHECK_OVERRIDE(hipModuleGetFunction(&func, mod, funcName.c_str()));
    return func;
}

static inline void launchKernel(
        hipFunction_t func,
        int gridX,
        int gridY,
        int gridZ,
        int numWraps,
        int sharedMemBytes,
        void* args[],
        int device_index) {
    AT_CUDA_DRIVER_CHECK_OVERRIDE(hipModuleLaunchKernel(
        func, gridX, gridY, gridZ, 32*numWraps, 1, 1, sharedMemBytes,
        at::hip::getCurrentHIPStreamMasqueradingAsCUDA(device_index), args, nullptr));
}

static hipFunction_t triton_unk_fused_addmm_0 = nullptr;

std::vector<at::Tensor> inductor_entry_cpp(const std::vector<at::Tensor>& args) {
    at::Tensor arg0_1 = args[0];
    at::Tensor arg1_1 = args[1];
    at::Tensor arg2_1 = args[2];

    c10::optional<at::Scalar> optional_scalar;
    c10::optional<c10::string_view> optional_string;
    torch::List<c10::optional<at::Scalar>> optional_list;

        at::hip::HIPGuardMasqueradingAsCUDA device_guard(0);
        auto buf0 = at::empty_strided({20L, 16L}, {16L, 1L}, at::device(c10::Device(at::kCUDA, 0)).dtype(at::kFloat));
        if (triton_unk_fused_addmm_0 == nullptr) {
             triton_unk_fused_addmm_0 = loadKernel("/tmp/torchinductor_root/5k/c5kxlv75iixqttomxk3ngp7iqyyomvkxcxh2yhod5lzfoem5cgcs.hsaco", "triton_unk_fused_addmm_0_0d1d2d3d");
        }
        hipDeviceptr_t var_0 = reinterpret_cast<hipDeviceptr_t>(arg2_1.data_ptr());
        hipDeviceptr_t var_1 = reinterpret_cast<hipDeviceptr_t>(arg0_1.data_ptr());
        hipDeviceptr_t var_2 = reinterpret_cast<hipDeviceptr_t>(arg1_1.data_ptr());
        hipDeviceptr_t var_3 = reinterpret_cast<hipDeviceptr_t>(buf0.data_ptr());
        void* kernel_args_var_0[] = {&var_0, &var_1, &var_2, &var_3};
        launchKernel(triton_unk_fused_addmm_0, 1, 1, 1, 2, 12288, kernel_args_var_0, 0);
        arg0_1.reset();
        arg1_1.reset();
        arg2_1.reset();
        return {buf0};
}
'''
)

module = CppWrapperCodeCache.load(cpp_wrapper_src, 'inductor_entry_cpp', 'cctownpl2aldbxbm6hosyfaoj5ij6n4sv32bembaek5gs3omqkdb', True)

def _wrap_func(f):
    def g(args):
        args_tensor = [arg if isinstance(arg, torch.Tensor) else torch.tensor(arg) for arg in args]
        return f(args_tensor)
    return g
call = _wrap_func(module.inductor_entry_cpp)


def benchmark_compiled_module(times=10, repeat=10):
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((20, 33), (33, 1), device='cuda:0', dtype=torch.float32)
    arg1_1 = rand_strided((33, 16), (16, 1), device='cuda:0', dtype=torch.float32)
    arg2_1 = rand_strided((20, 16), (16, 1), device='cuda:0', dtype=torch.float32)
    return print_performance(lambda: call([arg0_1, arg1_1, arg2_1]), times=times, repeat=repeat)


if __name__ == "__main__":
    from torch._inductor.utils import compiled_module_main
    compiled_module_main('None', benchmark_compiled_module)
```

**Comment by jataylo** [comment]
Pytorch bot doesn't like me trying to rebase, will resolve locally and force push

**Comment by jataylo** [comment]
```
2023-08-25T16:51:13.9953434Z - generated xml file: /var/lib/jenkins/pytorch/test/test-reports/python-pytest/inductor.test_cpp_wrapper/inductor.test_cpp_wrapper-7bfc89039bba36d5.xml -
2023-08-25T16:51:13.9954063Z =========== 115 passed, 36 skipped, 1 xfailed in 3318.56s (0:55:18) ============
```

Finally green on this opening up for review to add cpp wrapper support on ROCm with hsaco binaries cc: @jansel @eellison

**Comment by jataylo** [comment]
@jithunnair-amd do you know why the ROCm trunk jobs are being skipped?

**Comment by jataylo** [comment]
Rebasing to try and kick off ROCM ci

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -809,10 +809,13 @@ def get_include_and_linking_paths(
             else:
                 macros = f"-D{macros}"
         if cuda:
-            if config.is_fbcode():
-                libs += ["cuda"]
```
Comment (jansel) [inline torch/_inductor/codecache.py:813]: These lines are likely needed, and won't be covered by github tests.

Comment (jataylo) [reply torch/_inductor/codecache.py:813]: Thanks for taking a look @jansel, we haven't removed these for nvidia but we have conditionalised them out for ROCm only replacing with the equivalent rocm/hip libraries.

Full snippet
```
if cuda:
    if torch.version.hip is not None:
        libs += ["c10_hip", "torch_hip"]
    else:
        if config.is_fbcode():
            libs += ["cuda"]
        else:
            libs += ["c10_cuda", "cuda", "torch_cuda"]
```

Comment (jansel) [reply torch/_inductor/codecache.py:813]: Got it, my mistake

**Review by jansel — APPROVED** [review]

**Comment by jataylo** [comment]
ROCm tests are passing, failures in unstable are unrelated to the change. Will attempt to merge

**Comment by jataylo** [comment]
cc: @pragupta

**Comment by DanilBaibak** [comment]
@jansel unfortunately, I have to revert your PR because it breaks the internal build. The problem comes from [this change](https://github.com/pytorch/pytorch/pull/105141/files#diff-76e100027ee5abf25c7802a37ab3035802522ff4fdb1ddfbc1bd62b6f1306743R198). The file exists, but the internal build system cannot determine the path correctly. I'm trying to figure out how to fix this.

**Comment by jataylo** [comment]
@kit1980 @DanilBaibak Hi both I was wondering if there was any update regarding this so we can enable our cpp_wrapper support?

**Comment by jataylo** [comment]
Seeing new failures after getting this PR updated.

Performing local triaging.

**Comment by jataylo** [comment]
Green on cpp_wrapper and rocm tests now but seems to be errors in CUDA's codecache UTs.

Not sure if this issue comes from our PR or not yet.

**Comment by jataylo** [comment]
@DanilBaibak @kit1980 the remaining failures are unrelated. Could we get your help with merging this again with any steps required for the internal builds.

File: torch/_inductor/codecache.py [hunk 2]
```diff
@@ -346,7 +346,7 @@ def get_path(
 def get_hash(content: Union[str, bytes], extra: str = "", hash_type: str = "code"):
     if hash_type == "code":
         return code_hash(content, extra)
-    if hash_type == "cubin":
+    if hash_type == "cubin" or "hsaco":
```
Comment (malfet) [inline torch/_inductor/codecache.py:349]: Sorry, but `hash_type == "cubin" or "hsaco"` is equivalent to `hash_type == "cubin"`, isn't it?
```suggestion
    if hash_type in ["cubin", "hsaco"]:
```

File: torch/_inductor/codecache.py [hunk 3]
```diff
@@ -1502,15 +1505,21 @@ class CudaKernelParamCache:
...
             specified_dir=split_aot_inductor_output_path(
                 config.aot_inductor.output_path
             )[0],
         )
-        params["cubin_path"] = path
+
+        if torch.version.hip is None:
+            params["cubin_path"] = path
+        else:
+            params["hsaco_path"] = path
```
Comment (malfet) [inline torch/_inductor/codecache.py:1521]: Nit
```suggestion
            params["cubin_path" if torch.version.hip is None else "hsaco_path"] = path
```

File: torch/_inductor/codegen/wrapper.py [hunk 4]
```diff
@@ -2582,7 +2582,9 @@ def generate_kernel_call(
         ), f"cuda kernel parameters for {name} should already exist at this moment"
         mangled_name = params.get("mangled_name", None)
         assert mangled_name is not None, "missing mangled_name"
-        cubin_path = params.get("cubin_path", None)
+        cubin_path = params.get(
+            "cubin_path" if torch.version.hip is None else "hsaco_path", None
```
Comment (malfet) [inline torch/_inductor/codegen/wrapper.py:2586]: Considering that you need exactly the same in codecache, I wonder if we can turn it into a helper function... (like `get_cubin_path_name`)

**Review by malfet — CHANGES_REQUESTED** [review]
Please fix `or "hsaco_path"` bug

Also, please explain why changes to `cuda_to_hip_mappings` are needed in this PR

File: torch/_inductor/codegen/wrapper.py (hunk 4 above)
Comment (jataylo) [reply torch/_inductor/codegen/wrapper.py:2586]: Seems sensible looks like I'm thinking `cpp_wrapper_cubin_path` inline with the `cpp_wrapper_cache_dir` naming here: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codecache.py#L114

I'll update the PR with all suggested changes shortly

File: torch/_inductor/codecache.py (hunk 3 above)
Comment (jataylo) [reply torch/_inductor/codecache.py]: Adopted suggestion, thank you!

File: torch/_inductor/codegen/wrapper.py (hunk 4 above)
Comment (jataylo) [reply torch/_inductor/codegen/wrapper.py]: Adopted this suggestion. @malfet does this look okay?
[get_cpp_wrapper_cubin_path_name](https://github.com/pytorch/pytorch/blob/6749d0208958c86a0057bca7ca1e7a73bb2a404d/torch/_inductor/codecache.py#L129C5-L129C36)

```
def get_cpp_wrapper_cubin_path_name():
    return "cubin_path" if torch.version.hip is None else "hsaco_path"
```

Comment (jataylo) [reply torch/_inductor/codegen/wrapper.py]: @malfet In regards to `cuda_to_hip_mappings` change.

This feature will generate a C++ file which uses CUDA to launch triton kernels from cubin binaries. To support this on ROCm we are hipifying the actual codegen logic in wrapper.py to generate an equivalent HIP cpp wrapper code.

Here is a gist with the hipified file contents:
https://gist.github.com/jataylo/df6cf593e204e5196b64c7e47e056083

**Comment by jataylo** [comment]
@malfet Thank you for the review, I have updated this PR. Could you please help us take another look?

Note I believe we still need help from the likes of @DanilBaibak @kit1980 to help us merge this without breaking internal builds, last time we merged it had to be reverted.

**Comment by jataylo** [comment]
cc: @malfet rerunning CI without the `keep-going` label to avoid unrelated failures from https://github.com/pytorch/pytorch/issues/113419

File: tools/amd_build/build_amd.py [hunk 5]
```diff
@@ -193,6 +193,7 @@ def remove_hcc(line: str) -> str:
     output_directory=out_dir,
     includes=includes,
     ignores=ignores,
+    extra_files=["torch/_inductor/codegen/wrapper.py"],
```
Comment (malfet) [inline tools/amd_build/build_amd.py:196]: All internal builds are failing due to this line.

Comment (malfet) [reply tools/amd_build/build_amd.py:196]: Ok, found a workaround by providing a dummy to fb_caffe2_hipify.sh, but need to think of a proper fix. cc: @xw285cornell @bertmaher

**Review by malfet — APPROVED** [review]
